In [1]:
# pip install mlxtend
!pip install mlxtend
#(mlxtend --> machine learning extension)

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
olr = pd.read_excel("/content/Online retail.xlsx",header=None)
olr

0
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7498                                            chicken
7499                                 escalope,green tea
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7501 rows x 1 columns]

# Data exploration and Cleaning

In [3]:
olr.shape

(7501, 1)

In [4]:
olr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       7501 non-null   object
dtypes: object(1)
memory usage: 58.7+ KB


In [5]:
olr.dropna(inplace = True)

In [6]:
olr.drop_duplicates(inplace = True)

In [7]:
olr.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5176 entries, 0 to 7500
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       5176 non-null   object
dtypes: object(1)
memory usage: 80.9+ KB


In [8]:
olr.columns

Index([0], dtype='int64')

In [9]:
value_counts = olr[0:19].apply(pd.Series.value_counts)
value_counts

0
0                                                    
shrimp,almonds,avocado,vegetables mix,green gra...  1
burgers,meatballs,eggs                              1
chutney                                             1
turkey,avocado                                      1
mineral water,milk,energy bar,whole wheat rice,...  1
low fat yogurt                                      1
whole wheat pasta,french fries                      1
soup,light cream,shallot                            1
frozen vegetables,spaghetti,green tea               1
french fries                                        1
eggs,pet food                                       1
cookies                                             1
turkey,burgers,mineral water,eggs,cooking oil       1
spaghetti,champagne,cookies                         1
mineral water,salmon                                1
mineral water                                       1
shrimp,chocolate,chicken,honey,oil,cooking oil,...  1
turkey,eggs                                         1
turkey,fresh tuna,tomatoes,spaghetti,mineral wa...  1

# Association Rule mining

is data mining technique used for the large dataset. It involves finding relationship between variables in the large dataset

In [10]:
transactions = []
for i in range(len(olr)):
    transactions.append(olr.iloc[i, 0].split(','))

olr = pd.DataFrame(transactions).T

In [11]:
transactions

[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
 ['low fat yogurt'],
 ['whole wheat pasta', 'french fries'],
 ['soup', 'light cream', 'shallot'],
 ['frozen vegetables', 'spaghetti', 'green tea'],
 ['french fries'],
 ['eggs', 'pet food'],
 ['cookies'],
 ['turkey', 'burgers', 'mineral water', 'eggs', 'cooking oil'],
 ['spaghetti', 'champagne', 'cookies'],
 ['mineral water', 'salmon'],
 ['mineral water'],
 ['shrimp',
  'chocolate',
  'chicken',
  'honey',
  'oil',
  'cooking oil',
  'low fat yogurt'],
 ['turkey', 'eggs'],
 ['turkey',
  'fresh tuna',
  'tomatoes',
  'spagh

In [12]:
olr.shape

(20, 5176)

In [13]:
# import libraries for mlxtend
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

# Transaction encoder will convert dataset into True or False
from mlxtend.preprocessing import TransactionEncoder

In [14]:
# transaction encoder
# convert transaction of list to columns using TransactionEncoder
trans_encoder = TransactionEncoder()
trans_ary = trans_encoder.fit_transform(transactions)
# trans_encoder = encoder
# trans_ary = one_hot
# create a new dataframe from trans_ary with the column names
olr1 = pd.DataFrame(trans_ary, columns = trans_encoder.columns_)
olr1

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0          False     True               True      False     True        False   
1          False    False              False      False    False        False   
2          False    False              False      False    False        False   
3          False    False              False      False     True        False   
4          False    False              False      False    False        False   
...          ...      ...                ...        ...      ...          ...   
5171       False    False              False      False    False        False   
5172       False    False              False      False    False        False   
5173       False    False              False      False    False        False   
5174       False    False              False      False    False        False   
5175       False    False              False      False    False        False   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0     False           False      False        False  ...   False   
1     False           False      False        False  ...   False   
2     False           False      False        False  ...   False   
3     False           False      False        False  ...    True   
4     False           False      False        False  ...   False   
...     ...             ...        ...          ...  ...     ...   
5171  False           False      False        False  ...   False   
5172  False           False      False        False  ...    True   
5173  False           False      False        False  ...   False   
5174  False           False      False        False  ...   False   
5175  False           False      False        False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0               True        False       False              True   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
5171           False        False       False             False   
5172           False        False       False             False   
5173           False        False       False             False   
5174           False        False       False             False   
5175           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False   True        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False             False  False        False     False  
4                 False              True  False        False     False  
...                 ...               ...    ...          ...       ...  
5171              False             False  False        False     False  
5172              False              True  False        False     False  
5173              False             False  False        False     False  
5174              False             False  False        False     False  
5175              False             False  False         True     False  

[5176 rows x 120 columns]

In [15]:
trans_ary

array([[False,  True,  True, ...,  True, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False,  True, False]])

In [16]:
# convert dataset into 1-0 encoding
def encode_units(x):
    if x== False:
        return 0
    if x== True:
        return 1

In [17]:
# .applymap will map the data True to 1 & False to 0(by calling the function encode_units)
olr1 = olr1.applymap(encode_units)
olr1
# olr1 -> dataframe which contains the transaction data

/tmp/ipython-input-17-1014245253.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  olr1 = olr1.applymap(encode_units)


asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0              0        1                  1          0        1            0   
1              0        0                  0          0        0            0   
2              0        0                  0          0        0            0   
3              0        0                  0          0        1            0   
4              0        0                  0          0        0            0   
...          ...      ...                ...        ...      ...          ...   
5171           0        0                  0          0        0            0   
5172           0        0                  0          0        0            0   
5173           0        0                  0          0        0            0   
5174           0        0                  0          0        0            0   
5175           0        0                  0          0        0            0   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0         0               0          0            0  ...       0   
1         0               0          0            0  ...       0   
2         0               0          0            0  ...       0   
3         0               0          0            0  ...       1   
4         0               0          0            0  ...       0   
...     ...             ...        ...          ...  ...     ...   
5171      0               0          0            0  ...       0   
5172      0               0          0            0  ...       1   
5173      0               0          0            0  ...       0   
5174      0               0          0            0  ...       0   
5175      0               0          0            0  ...       0   

      vegetables mix  water spray  white wine  whole weat flour  \
0                  1            0           0                 1   
1                  0            0           0                 0   
2                  0            0           0                 0   
3                  0            0           0                 0   
4                  0            0           0                 0   
...              ...          ...         ...               ...   
5171               0            0           0                 0   
5172               0            0           0                 0   
5173               0            0           0                 0   
5174               0            0           0                 0   
5175               0            0           0                 0   

      whole wheat pasta  whole wheat rice  yams  yogurt cake  zucchini  
0                     0                 0     1            0         0  
1                     0                 0     0            0         0  
2                     0                 0     0            0         0  
3                     0                 0     0            0         0  
4                     0                 1     0            0         0  
...                 ...               ...   ...          ...       ...  
5171                  0                 0     0            0         0  
5172                  0                 1     0            0         0  
5173                  0                 0     0            0         0  
5174                  0                 0     0            0         0  
5175                  0                 0     0            1         0  

[5176 rows x 120 columns]

In [18]:
# Implement an Apriori algorithm for itemset(data) mining and association rule learning
frequent_itemsets = apriori(olr1, min_support = 0.02, use_colnames = True)
frequent_itemsets
# frequent_itemsets contains all the itemsets which has consider the minimum support threshold = 0.02
# min_support = 0.02 means that it consider the itemsets appeared in at least 2% of transactions.

/usr/local/lib/python3.11/dist-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                 itemsets
0    0.029366                                (almonds)
1    0.045981                                (avocado)
2    0.020479                              (black tea)
3    0.045015                               (brownies)
4    0.113794                                (burgers)
..        ...                                      ...
169  0.020093         (chocolate, milk, mineral water)
170  0.022991    (spaghetti, chocolate, mineral water)
171  0.020672         (spaghetti, eggs, mineral water)
172  0.024730  (spaghetti, ground beef, mineral water)
173  0.022604         (spaghetti, milk, mineral water)

[174 rows x 2 columns]

In [19]:
# Apply association rule mining techniques to the pre-processed dataset
#to discver intersting relationship b/w  products that purchased together
rules = association_rules(frequent_itemsets, metric = "confidence", min_threshold = 0.4)
rules

# metric = "confidence" -> sets the metric to be used for generating association rules.
# min_threshold = 0.4-> rule with confidence of atleats 40% will be considered.

antecedents      consequents  antecedent support  \
0                  (cooking oil)  (mineral water)            0.071870   
1                  (ground beef)  (mineral water)            0.135819   
2                  (ground beef)      (spaghetti)            0.135819   
3                    (olive oil)  (mineral water)            0.087713   
4                         (soup)  (mineral water)            0.070904   
5              (chocolate, milk)  (mineral water)            0.044822   
6         (chocolate, spaghetti)  (mineral water)            0.055835   
7              (eggs, spaghetti)  (mineral water)            0.051391   
8       (ground beef, spaghetti)  (mineral water)            0.055835   
9   (ground beef, mineral water)      (spaghetti)            0.058733   
10             (milk, spaghetti)  (mineral water)            0.050039   

    consequent support   support  confidence      lift  representativity  \
0             0.299845  0.028980    0.403226  1.344779               1.0   
1             0.299845  0.058733    0.432432  1.442184               1.0   
2             0.229521  0.055835    0.411095  1.791102               1.0   
3             0.299845  0.038640    0.440529  1.469186               1.0   
4             0.299845  0.033423    0.471390  1.572109               1.0   
5             0.299845  0.020093    0.448276  1.495023               1.0   
6             0.299845  0.022991    0.411765  1.373257               1.0   
7             0.299845  0.020672    0.402256  1.341543               1.0   
8             0.299845  0.024730    0.442907  1.477116               1.0   
9             0.229521  0.024730    0.421053  1.834485               1.0   
10            0.299845  0.022604    0.451737  1.506568               1.0   

    leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0   0.007430    1.173232       0.276236  0.084555   0.147654    0.249938  
1   0.018008    1.233606       0.354795  0.155818   0.189368    0.314154  
2   0.024661    1.308326       0.511102  0.180400   0.235664    0.327181  
3   0.012340    1.251457       0.350055  0.110742   0.200932    0.284697  
4   0.012163    1.324519       0.391684  0.099084   0.245009    0.291429  
5   0.006653    1.269030       0.346652  0.061905   0.211997    0.257643  
6   0.006249    1.190263       0.287877  0.069106   0.159849    0.244220  
7   0.005263    1.171328       0.268382  0.062537   0.146268    0.235599  
8   0.007988    1.256799       0.342107  0.074723   0.204328    0.262690  
9   0.011249    1.330828       0.483272  0.093842   0.248588    0.264398  
10  0.007600    1.277042       0.353951  0.069067   0.216941    0.263562

# Lift
In the above dataset the lift values are having a positive association.
Bcz lift values greater than 1 indicate a +ve association and higher values.
And common starting point is around 1.2 to 3

In [20]:
'''
rules = association_rules(frequent_itemsets, metric = "confidence", min_threshold = 0.4)
rules = rules[rules['lift'] > 1.4]
rules
'''

'\nrules = association_rules(frequent_itemsets, metric = "confidence", min_threshold = 0.4) \nrules = rules[rules[\'lift\'] > 1.4]\nrules\n'

In [21]:
# sort by support
sorted_rules_support = rules.sort_values(by = "support", ascending = False)
sorted_rules_support

antecedents      consequents  antecedent support  \
1                  (ground beef)  (mineral water)            0.135819   
2                  (ground beef)      (spaghetti)            0.135819   
3                    (olive oil)  (mineral water)            0.087713   
4                         (soup)  (mineral water)            0.070904   
0                  (cooking oil)  (mineral water)            0.071870   
9   (ground beef, mineral water)      (spaghetti)            0.058733   
8       (ground beef, spaghetti)  (mineral water)            0.055835   
6         (chocolate, spaghetti)  (mineral water)            0.055835   
10             (milk, spaghetti)  (mineral water)            0.050039   
7              (eggs, spaghetti)  (mineral water)            0.051391   
5              (chocolate, milk)  (mineral water)            0.044822   

    consequent support   support  confidence      lift  representativity  \
1             0.299845  0.058733    0.432432  1.442184               1.0   
2             0.229521  0.055835    0.411095  1.791102               1.0   
3             0.299845  0.038640    0.440529  1.469186               1.0   
4             0.299845  0.033423    0.471390  1.572109               1.0   
0             0.299845  0.028980    0.403226  1.344779               1.0   
9             0.229521  0.024730    0.421053  1.834485               1.0   
8             0.299845  0.024730    0.442907  1.477116               1.0   
6             0.299845  0.022991    0.411765  1.373257               1.0   
10            0.299845  0.022604    0.451737  1.506568               1.0   
7             0.299845  0.020672    0.402256  1.341543               1.0   
5             0.299845  0.020093    0.448276  1.495023               1.0   

    leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
1   0.018008    1.233606       0.354795  0.155818   0.189368    0.314154  
2   0.024661    1.308326       0.511102  0.180400   0.235664    0.327181  
3   0.012340    1.251457       0.350055  0.110742   0.200932    0.284697  
4   0.012163    1.324519       0.391684  0.099084   0.245009    0.291429  
0   0.007430    1.173232       0.276236  0.084555   0.147654    0.249938  
9   0.011249    1.330828       0.483272  0.093842   0.248588    0.264398  
8   0.007988    1.256799       0.342107  0.074723   0.204328    0.262690  
6   0.006249    1.190263       0.287877  0.069106   0.159849    0.244220  
10  0.007600    1.277042       0.353951  0.069067   0.216941    0.263562  
7   0.005263    1.171328       0.268382  0.062537   0.146268    0.235599  
5   0.006653    1.269030       0.346652  0.061905   0.211997    0.257643

In [22]:
# sort by confidence
sorted_rules_confidence = rules.sort_values(by = "confidence", ascending=False)
sorted_rules_confidence

antecedents      consequents  antecedent support  \
4                         (soup)  (mineral water)            0.070904   
10             (milk, spaghetti)  (mineral water)            0.050039   
5              (chocolate, milk)  (mineral water)            0.044822   
8       (ground beef, spaghetti)  (mineral water)            0.055835   
3                    (olive oil)  (mineral water)            0.087713   
1                  (ground beef)  (mineral water)            0.135819   
9   (ground beef, mineral water)      (spaghetti)            0.058733   
6         (chocolate, spaghetti)  (mineral water)            0.055835   
2                  (ground beef)      (spaghetti)            0.135819   
0                  (cooking oil)  (mineral water)            0.071870   
7              (eggs, spaghetti)  (mineral water)            0.051391   

    consequent support   support  confidence      lift  representativity  \
4             0.299845  0.033423    0.471390  1.572109               1.0   
10            0.299845  0.022604    0.451737  1.506568               1.0   
5             0.299845  0.020093    0.448276  1.495023               1.0   
8             0.299845  0.024730    0.442907  1.477116               1.0   
3             0.299845  0.038640    0.440529  1.469186               1.0   
1             0.299845  0.058733    0.432432  1.442184               1.0   
9             0.229521  0.024730    0.421053  1.834485               1.0   
6             0.299845  0.022991    0.411765  1.373257               1.0   
2             0.229521  0.055835    0.411095  1.791102               1.0   
0             0.299845  0.028980    0.403226  1.344779               1.0   
7             0.299845  0.020672    0.402256  1.341543               1.0   

    leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
4   0.012163    1.324519       0.391684  0.099084   0.245009    0.291429  
10  0.007600    1.277042       0.353951  0.069067   0.216941    0.263562  
5   0.006653    1.269030       0.346652  0.061905   0.211997    0.257643  
8   0.007988    1.256799       0.342107  0.074723   0.204328    0.262690  
3   0.012340    1.251457       0.350055  0.110742   0.200932    0.284697  
1   0.018008    1.233606       0.354795  0.155818   0.189368    0.314154  
9   0.011249    1.330828       0.483272  0.093842   0.248588    0.264398  
6   0.006249    1.190263       0.287877  0.069106   0.159849    0.244220  
2   0.024661    1.308326       0.511102  0.180400   0.235664    0.327181  
0   0.007430    1.173232       0.276236  0.084555   0.147654    0.249938  
7   0.005263    1.171328       0.268382  0.062537   0.146268    0.235599

In [23]:
#sort by lift
sorted_rule_lift = rules.sort_values(by = 'lift', ascending = False)
sorted_rule_lift

antecedents      consequents  antecedent support  \
9   (ground beef, mineral water)      (spaghetti)            0.058733   
2                  (ground beef)      (spaghetti)            0.135819   
4                         (soup)  (mineral water)            0.070904   
10             (milk, spaghetti)  (mineral water)            0.050039   
5              (chocolate, milk)  (mineral water)            0.044822   
8       (ground beef, spaghetti)  (mineral water)            0.055835   
3                    (olive oil)  (mineral water)            0.087713   
1                  (ground beef)  (mineral water)            0.135819   
6         (chocolate, spaghetti)  (mineral water)            0.055835   
0                  (cooking oil)  (mineral water)            0.071870   
7              (eggs, spaghetti)  (mineral water)            0.051391   

    consequent support   support  confidence      lift  representativity  \
9             0.229521  0.024730    0.421053  1.834485               1.0   
2             0.229521  0.055835    0.411095  1.791102               1.0   
4             0.299845  0.033423    0.471390  1.572109               1.0   
10            0.299845  0.022604    0.451737  1.506568               1.0   
5             0.299845  0.020093    0.448276  1.495023               1.0   
8             0.299845  0.024730    0.442907  1.477116               1.0   
3             0.299845  0.038640    0.440529  1.469186               1.0   
1             0.299845  0.058733    0.432432  1.442184               1.0   
6             0.299845  0.022991    0.411765  1.373257               1.0   
0             0.299845  0.028980    0.403226  1.344779               1.0   
7             0.299845  0.020672    0.402256  1.341543               1.0   

    leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
9   0.011249    1.330828       0.483272  0.093842   0.248588    0.264398  
2   0.024661    1.308326       0.511102  0.180400   0.235664    0.327181  
4   0.012163    1.324519       0.391684  0.099084   0.245009    0.291429  
10  0.007600    1.277042       0.353951  0.069067   0.216941    0.263562  
5   0.006653    1.269030       0.346652  0.061905   0.211997    0.257643  
8   0.007988    1.256799       0.342107  0.074723   0.204328    0.262690  
3   0.012340    1.251457       0.350055  0.110742   0.200932    0.284697  
1   0.018008    1.233606       0.354795  0.155818   0.189368    0.314154  
6   0.006249    1.190263       0.287877  0.069106   0.159849    0.244220  
0   0.007430    1.173232       0.276236  0.084555   0.147654    0.249938  
7   0.005263    1.171328       0.268382  0.062537   0.146268    0.235599

# Analysis and Interpretation

support:
itemset (ground beef & mineral water) has high support which means that these itemsets are frequently purchased together by customers

confidence:
(soup & mineral water) has high confidence with 0.047 which means 40% transactions is considered in the dataset.

lift:
(ground beef,mineral water & spaghetti) has high lift values with 1.83 gives strong relationship b/w products .
And also these products are often purchased together.

# Interview Questions

# 1.What is lift and why is it important in Association rules?

Lift is a key metric used ina association rule mining to evaluate the strength and significance of an association.
Lift is used to measure the strength of association b/w to items.
Lift>1 indicates positive association, means A & B are occur together more often than expected by chance.
Lift<1 indicates negative association, means A & B are occur together less often than exoected chance.

# 2.What is support and Confidence. How do you calculate them?

Support: support of an itemset is the proportion of transaction in the dataset.
It measures how frequently an itemset occurs int the dataset.
    
   Support(A) = Number of  transactions containing A)/(Total number of transactions)

Confidence: measure the likelihood that item B is purchased when item A is purchased.

   Confidence(A->B) = Support(A ∩ B)/Supoort(A)

# 3.	What are some limitations or challenges of Association rules mining?
Low support:
patterns can be missed if minimum support thresholds are too high even though the items are Rare.

Difficulty in setting Thresholds:
choosing an appropriate support and confidence thresholds is subjective and may contain error.

Interpretation Challenges:
understanding and interpreting the rules, with complex relationship can be difficult.
